middleware.go

```diff
@@ -30,14 +69,248 @@ func LoggingMiddleware(log *logrus.Logger) func(http.Handler) http.Handler {
 			request.Latency = fmt.Sprintf("%.9fs", m.Duration.Seconds())
 			request.ResponseSize = strconv.FormatInt(m.Written, 10)
 
-			fields := logrus.Fields{"httpRequest": request}
-
-			traceHeader := r.Header.Get("X-Cloud-Trace-Context")
-			if traceHeader != "" {
-				fields["trace"] = traceHeader
+			if o.filterHTTP(r) {
+				// log the result
+				ctxlogrus.Extract(ctx).Infof("served HTTP %v %v", r.Method, r.URL)
 			}
+		})
+	}
+}
+
+func UnaryLoggingInterceptor(logger *logrus.Logger, opts ...MiddlewareOption) grpc.UnaryServerInterceptor {
+	o := evaluateMiddlewareOptions(opts)
+	return loggingInterceptor{logger: logger, middlewareOptions: o}.intercept
+}
+
+func StreamLoggingInterceptor(logger *logrus.Logger, opts ...MiddlewareOption) grpc.StreamServerInterceptor {
+	o := evaluateMiddlewareOptions(opts)
+	return loggingInterceptor{logger: logger, middlewareOptions: o}.interceptStream
+}
+
+type loggingInterceptor struct {
+	logger *logrus.Logger
+	*middlewareOptions
+}
+
+type GRPCRequest struct {
+	Method    string `json:"method,omitempty"`
+	UserAgent string `json:"userAgent,omitempty"`
+	PeerAddr  string `json:"peer,omitempty"`
+	Deadline  string `json:"deadline,omitempty"`
+	Duration  string `json:"duration,omitempty"`
+}
+
+func (l loggingInterceptor) intercept(
+	ctx context.Context,
+	req interface{},
+	info *grpc.UnaryServerInfo,
+	handler grpc.UnaryHandler,
+) (resp interface{}, err error) {
+	startTime := time.Now()
+	ctx = ctxlogrus.ToContext(ctx, logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	resp, err = handler(ctx, req)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
+
+	l.log(ctx, err, info.FullMethod)
+
+	return
+}
+
+func (l loggingInterceptor) interceptStream(
+	srv interface{},
+	ss grpc.ServerStream,
+	info *grpc.StreamServerInfo,
+	handler grpc.StreamHandler,
+) (err error) {
+	startTime := time.Now()
+	ctx := ctxlogrus.ToContext(ss.Context(), logrus.NewEntry(l.logger))
+
+	request := l.requestFromContext(ctx, &GRPCRequest{
+		Method: info.FullMethod,
+	})
+
+	wrapped := grpc_middleware.WrapServerStream(ss)
+	wrapped.WrappedContext = ctx
+
+	err = handler(srv, wrapped)
+
+	request.Duration = fmt.Sprintf("%.9fs", time.Since(startTime).Seconds())
```

Comment:
milliseconds is better I think

Reply:
See above re: the duration format prescribed by the protocol I'm copying here. If it's ever parsed and formatted for display, it'll need to be as Google defines, duration in seconds with up to 9 fractional digits terminated b `s`.